From 02f7cb1fa058b99b50e1d5cfe01d81117a653d15 Mon Sep 17 00:00:00 2001
From: Daniel Larraz <daniel-larraz@users.noreply.github.com>
Date: Wed, 28 Aug 2024 08:31:37 -0500
Subject: [PATCH 041/312] cmake: Check whether CoCoA has been patched (#11183)

If CoCoA has not been patched and the `--auto-download` option is
provided, CMake will build a patched version for the user.

Resolves #11182
---
 cmake/FindCoCoA.cmake           | 18 ++++++++++++++++++
 cmake/deps-utils/cocoa-test.cpp | 23 +++++++++++++++++++++++
 2 files changed, 41 insertions(+)
 create mode 100644 cmake/deps-utils/cocoa-test.cpp

diff --git a/cmake/FindCoCoA.cmake b/cmake/FindCoCoA.cmake
index 1783afe78..e8c8d0d14 100644
--- a/cmake/FindCoCoA.cmake
+++ b/cmake/FindCoCoA.cmake
@@ -31,6 +31,24 @@ if(CoCoA_INCLUDE_DIR AND CoCoA_LIBRARIES)
   string(REGEX MATCH "[0-9]+\\.[0-9]+" CoCoA_VERSION "${CoCoA_VERSION}")
 
   check_system_version("CoCoA")
+
+  # This test checks whether CoCoA has been patched
+  try_compile(CoCoA_USABLE "${DEPS_BASE}/try_compile/CoCoA-EP"
+    "${CMAKE_CURRENT_LIST_DIR}/deps-utils/cocoa-test.cpp"
+    CMAKE_FLAGS
+      "-DCMAKE_TOOLCHAIN_FILE=${CMAKE_TOOLCHAIN_FILE}"
+      "-DINCLUDE_DIRECTORIES=${CoCoA_INCLUDE_DIR}"
+    LINK_LIBRARIES ${CoCoA_LIBRARIES} ${GMP_LIBRARIES} ${GMPXX_LIBRARIES}
+  )
+  if(NOT CoCoA_USABLE)
+    message(STATUS "The system version of CoCoA has not been patched.")
+    if(ENABLE_AUTO_DOWNLOAD)
+      message(STATUS "A patched version of CoCoA will be built for you.")
+    else()
+      message(STATUS "Use --auto-download to let us download and build a patched version of CoCoA for you.")
+    endif()
+    set(CoCoA_FOUND_SYSTEM FALSE)
+  endif()
 endif()
 
 if(NOT CoCoA_FOUND_SYSTEM)
diff --git a/cmake/deps-utils/cocoa-test.cpp b/cmake/deps-utils/cocoa-test.cpp
new file mode 100644
index 000000000..94b6c7bd9
--- /dev/null
+++ b/cmake/deps-utils/cocoa-test.cpp
@@ -0,0 +1,23 @@
+/******************************************************************************
+ * Top contributors (to current version):
+ *   Daniel Larraz
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2024 by the authors listed in the file AUTHORS
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * Checks whether CoCoA has been patched.
+ */
+#include <iostream>
+#include <CoCoA/TmpGPoly.H>
+
+int main()
+{
+    std::cout << "CoCoA::handlersEnabled = ";
+    std::cout << CoCoA::handlersEnabled << std::endl;
+    return 0;
+}
\ No newline at end of file
-- 
2.47.1

